Ignore error of symlink

junjihashimoto opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
When setupCompilerEnvironmentPhase, symlink fails for existing link.
I want to ignore the error.
This pr changes `ln -s` to `ln -sf`.

OS : macos
```
{
  "url": "https://github.com/nixos/nixpkgs-channels.git",
  "rev": "3320a06049fc259e87a2bd98f4cd42f15f746b96",
  "date": "2020-03-26T20:57:21-04:00",
  "sha256": "1g5l186d5xh187vdcpfsz1ff8s749949c1pclvzfkylpar09ldkl",
  "fetchSubmodules": false
}
```

![image](https://user-images.githubusercontent.com/2469618/77827345-6a239900-7158-11ea-8f1b-1075bee4e4bc.png)
https://github.com/hasktorch/hasktorch/runs/541719980


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jtojnar committed:
haskellPackages.llvm-hs: Use LLVM 9

junjihashimoto committed:
Ignore error of symlink

cdepillabout commented:
This code was last changed by @Infinisil in https://github.com/NixOS/nixpkgs/commit/8d4509e34d85501dca2eed1a2dfbd5c0611b0f5b.

@Infinisil Is it possible that there was a bug introduced in https://github.com/NixOS/nixpkgs/commit/8d4509e34d85501dca2eed1a2dfbd5c0611b0f5b for when there are multiple `*.dylib` files that have the same name?

Is the fix that @junjihashimoto is proposing here correct?

nh2 commented:
If `-f` is added we definitely need a (commented) explanation of what happens when multiple sources provide the same file (the last one will win), and why that makes sense.

Is it correct that there _should_ be such conflicts in the first place?

junjihashimoto commented:
@nh2 
there are conf-files for each library(haskell's package) in $packageConfDir.
And dynamic-library-dirs is path of shared libraries the library uses.

For example, 
Assuming we have three libraries (A, B, C),
we will have A.conf, B.conf and C.conf in $packageConfDir.
When A depends on B, and B depends on C, and C has libC.so,
we will  take duplicate paths of libC.so.

```
    for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
      for lib in "$d/"*.{dylib,so}; do
        ln -s "$lib" "$dynamicLinksDir"
        ln -sf "$lib" "$dynamicLinksDir"
      done
    done
```

gnprice commented:
> Is it possible that there was a bug introduced in [8d4509e](https://github.com/NixOS/nixpkgs/commit/8d4509e34d85501dca2eed1a2dfbd5c0611b0f5b) for when there are multiple `*.dylib` files that have the same name?

I *think*, from reading that commit, that the code should try to create the same set of links before and after the commit. (Outside of cases of the bug that the commit was fixing.) If so, then this has probably behaved the same way since this `lib/links/` directory was introduced in #25537.


> When A depends on B, and B depends on C, and C has libC.so,
> we will take duplicate paths of libC.so.
> But the duplicated paths are all the same, so I think we can ignore this error.

If that's what's going on, then perhaps we can have the script check that: if the link already exists, it should point to the same path we were going to create it with. Then if it doesn't, we create it just like we do now.

chkno committed:
haskellPackages.idris: Fix build (new GHC 8.8 & old megaparsec 7)

The hackage-packages.nix change was generated by hackage2nix v2.15.1 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/ad4a70d448051948f2064d0bba5e63eab25f04d8.

junjihashimoto committed:
Remove duplicated links

junjihashimoto commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -383,11 +383,20 @@ stdenv.mkDerivation ({
+    listDynamicLinks () {
+      for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
+        for lib in "$d/"*.{dylib,so} ; do
+          if [ -f "$lib" ] ; then
```
When we use linux, lib includes not only actual filenames but also `*.dylib`.
This `[ -f "$lib" ]` is used to delete `*.dylib`.

gnprice replied:
Mmm, I see -- because there are no files matching `somedir/*.dylib`?

Hmm, but this bit of code has always been this way -- in fact it's only since February in 019e86fde that it mentions `.so` at all, and before that it was only `.dylib`:
```diff
     for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
-      for lib in "$d/"*.dylib; do
+      for lib in "$d/"*.{dylib,so}; do
         ln -s "$lib" "$dynamicLinksDir"
```

This change should be a separate commit, in any event. Probably deserves some separate debugging and discussion, too. It sounds like an independent issue from the collision issue.

junjihashimoto commented:
@gnprice 
I update this pr to to generating duplicate links.
Is it fine like this?

cdepillabout commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -383,11 +383,20 @@ stdenv.mkDerivation ({
-    for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
-      for lib in "$d/"*.{dylib,so}; do
-        ln -s "$lib" "$dynamicLinksDir"
+    listDynamicLinks () {
```
Could you add a comment on this function explaining what it is doing (and why it is needed)?

cdepillabout reviewed (changes requested):
I left one comment above.

Also, please rebase this on the `haskell-updates` branch.

gnprice commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -383,11 +383,20 @@ stdenv.mkDerivation ({
-        ln -s "$lib" "$dynamicLinksDir"
+    listDynamicLinks () {
+      for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
+        for lib in "$d/"*.{dylib,so} ; do
+          if [ -f "$lib" ] ; then
+            echo "$lib"
+          fi
+        done
```
I was thinking of something like:
```shell
  for lib in "$d/"*.{dylib,so} ; do
    local linkPath="$dynamicLinksDir/$(basename "$lib")"
    if [ -e "$linkPath" ]; then
      # link already exists! but maybe that's OK
      if [ ! -l "$linkPath" ] || [ "$(readlink "$linkPath")" != "$lib" ]; then
        echo >&2 "error: failed to create symbolic link $linkPath: file exists"
        exit 1
      fi
    else
      ln -s "$lib" "$linkPath"
    fi
  done
```

I.e., if the place where we want to create the link already exists, then check that it is already a link and already points exactly where we'd like to point it.

gnprice commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -383,11 +383,20 @@ stdenv.mkDerivation ({
       done
+    }
+
+    for lib in $(listDynamicLinks | sort | uniq) ; do
+      ln -s "$lib" "$dynamicLinksDir"
     done
```
I think the `ln -s` step can best be pushed inside the function. Then this is just a call, like:
```shell
    makeDynamicLinks
```
That just keeps the `ln -s` a little closer to the logic that drives it.

(Also, as is I think the `sort | uniq` shouldn't be needed, because of the `sort -u` that's inside the function.)

junjihashimoto replied:
`sort | uniq` deletes duplicated links.
`sort -u` deletes duplicated directories.
If we delete either, we should delete `sort -u`.

junjihashimoto replied:
Thank you for your reviewing.
I will update it and check the test.

gnprice reviewed (commented):
Thanks @junjihashimoto for the revisions! Comments below.

cdepillabout committed:
Merge pull request #84131 from chkno/idris

haskellPackages.idris: Fix build (new GHC 8.8 & old megaparsec 7)

turboMaCk committed:
haskellPackages.persistent-postgresql: don't check

Test suite requires database. Disabling tests as is done for mysql backend.

cdepillabout committed:
Merge pull request #84272 from turboMaCk/persistent-postgresql

haskellPackages.persistent-postgresql: don't check

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.15.1 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/0123dd80c207856fcec1db338417d30d73b36bf0.

peti committed:
LTS Haskell 15.7

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.15.1 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/8c7fc4a25f17b656a22232032fd4f3ebfd9a4889.

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.15.1 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/e07c49fea6de716fb6b06fe5293b2fd55b0692dd.

mikesperber committed:
haskellPackages.tensorflow: Update dependencies

This is the first step towards unbreaking the tensorflow packages.

adamse committed:
ghc: don't use ld.gold with musl libc (fixes #84670)

ld.gold doesn't play well with musl as is documented in #49071 and
https://sourceware.org/bugzilla/show_bug.cgi?id=23856

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.15.1 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/b4a552bd842c09fe8e762082c9c427e04090d10e.

cdepillabout committed:
Merge pull request #84720 from mikesperber/unbreak-tensorflow-haskell-1

Unbreak tensorflow haskell 1

cdepillabout committed:
Merge pull request #84741 from adamse/haskell-updates

ghc: do not use ld.gold with musl libc #84670

samdoshi committed:
haskellPackages.hsexif: unbroken

cdepillabout committed:
Merge pull request #84782 from samdoshi/hsexif

haskellPackages.hsexif: unbroken (fix

adamse committed:
ghc: mention why ld.gold is disabled for musl libc

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.15.1 from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/9501e1f4521de9b394d9c3ebafec3348122c528c.

cdepillabout committed:
Merge pull request #84809 from adamse/haskell-updates

ghc: mention why ld.gold is disabled for musl libc

junjihashimoto committed:
Add makeDynamicLinks and its comments

junjihashimoto commented:
> I left one comment above.
> 
> Also, please rebase this on the `haskell-updates` branch.

I rebase this on the `haskell-updates` branch, but conflicts happen.
Is it right?

cdepillabout commented:
@junjihashimoto 

You'll need to change the base branch on this PR to be `haskell-updates`, not `master`.

nh2 commented:
@junjihashimoto By the way, you do not need to make a new PR to change branches, you can click the `Edit` button here for that:

![image](https://user-images.githubusercontent.com/399535/79044614-69820c80-7c06-11ea-9a0c-68d47ee9f929.png)

nh2 commented:
For ticket subscribers, new PR is in #84985

junjihashimoto commented:
Thank you for letting me know.
